upgrade/src/main/java/org/odk/collect/upgrade/LaunchState.kt

```diff
@@ -0,0 +1,32 @@
+package org.odk.collect.upgrade
+
+import org.odk.collect.shared.Settings
+
+interface LaunchState {
+    fun isUpgradedFirstLaunch(): Boolean
+    fun appLaunched()
+}
+
+class VersionCodeLaunchState(
+    private val key: String,
+    private val settings: Settings,
+    private val currentVersion: Int,
+    private val installDetector: InstallDetector? = null
+) : LaunchState {
+
+    override fun isUpgradedFirstLaunch(): Boolean {
+        return if (settings.contains(key)) {
+            settings.getInt(key) < currentVersion
+        } else {
+            return installDetector?.installDetected() ?: false
+        }
+    }
+
+    override fun appLaunched() {
+        settings.save(key, currentVersion)
+    }
+}
+
+interface InstallDetector {
```

Comment:
Why did you put this interface here? I think we usually would do that in the same class with its implementation if it has only one implementation.

Reply:
The `upgrade` module owns the interface here - `LaunchState` and `BeforeProjectsInstallDetector` both depend on `InstallDetector`. In Java, I'd probably have given this its own file but I kind of like it just sitting beside the place that needs it in Kotlin?

Reply:
But if `BeforeProjectsInstallDetector` was in `upgrade` module you would probably put that interface there since it's the only implementation?
Is there any reason why BeforeProjectsInstallDetector is not there (in `upgrade` module) it seems like it is not coupled with any other modules so it could be I guess?

Reply:
It could be moved to `upgrade` given that there are no logical dependencies between `BeforeProjectsInstallDetector` and anything else in `collect_app`, but for me `BeforeProjectsInstallDetector` is "Collect specific" code whereas `upgrade` could work with any Android app(theoretically). I'd see it as a domain split. In other words, by exposing the `InstallDetector` interface `upgrade` is saying to `collect_app` "I need one of these passed to me - you choose how it works"

Reply:
ok if it's a matter of taste and not something important the one who implemented it (you in this case) can decide so let's leave what you like more.